File: src/reccobeats.py
"""ReccoBeats API client — audio features using Spotify track IDs."""

import time

import requests

BASE_URL = "https://api.reccobeats.com/v1"
BATCH_SIZE = 40
RATE_LIMIT_DELAY = 2.0

FEATURE_KEYS = [
    "acousticness", "danceability", "energy", "instrumentalness",
    "liveness", "loudness", "speechiness", "tempo", "valence",
]


def _extract_spotify_id(href: str) -> str:
    """Extract Spotify track ID from href like https://open.spotify.com/track/XXX."""
    if "/track/" in href:
        return href.split("/track/")[-1].split("?")[0]
    return ""


def get_audio_features_batch(track_ids: list[str]) -> dict[str, dict[str, float]]:
    """Fetch audio features for tracks by Spotify ID.

    Returns dict: spotify_track_id -> {acousticness, danceability, energy, ...}
    """
    results: dict[str, dict[str, float]] = {}

    for i in range(0, len(track_ids), BATCH_SIZE):
        batch = track_ids[i:i + BATCH_SIZE]
        ids_param = ",".join(batch)
        url = f"{BASE_URL}/audio-features?ids={ids_param}"

        try:
            resp = requests.get(url, headers={"Accept": "application/json"}, timeout=30)
            if resp.status_code == 200:
                data = resp.json()
                items = data.get("content", data if isinstance(data, list) else [])
                for item in items:
                    if not item:
                        continue
                    href = item.get("href", "")
                    spotify_id = _extract_spotify_id(href)
                    if spotify_id:
                        features = {k: float(item[k]) for k in FEATURE_KEYS if item.get(k) is not None}
                        if features:
                            results[spotify_id] = features
        except Exception as e:
            print(f"  ReccoBeats batch {i // BATCH_SIZE + 1} error: {e}")

        if i + BATCH_SIZE < len(track_ids):
            time.sleep(RATE_LIMIT_DELAY)

    return results

File: web/src/lib/reccobeats.ts
const BATCH_SIZE = 40;
const RATE_LIMIT_MS = 2000;

export const FEATURE_KEYS = [
  "acousticness", "danceability", "energy", "instrumentalness",
  "liveness", "loudness", "speechiness", "tempo", "valence",
] as const;

export type FeatureKey = (typeof FEATURE_KEYS)[number];
export type AudioFeatures = Record<FeatureKey, number>;

function extractSpotifyId(href: string): string {
  const match = href.match(/\/track\/([a-zA-Z0-9]+)/);
  return match?.[1] ?? "";
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function getAudioFeaturesBatch(
  trackIds: string[],
  onProgress?: (done: number, total: number) => void,
): Promise<Map<string, AudioFeatures>> {
  const results = new Map<string, AudioFeatures>();

  for (let i = 0; i < trackIds.length; i += BATCH_SIZE) {
    const batch = trackIds.slice(i, i + BATCH_SIZE);

    try {
      // [C1 FIX] Call local API route proxy, not ReccoBeats directly
      const resp = await fetch(`/api/reccobeats?ids=${batch.join(",")}`);
      if (resp.ok) {
        const data = await resp.json();
        const items: Array<Record<string, unknown>> = data.content ?? data ?? [];
        for (const item of items) {
          if (!item) continue;
          const spotifyId = extractSpotifyId(String(item.href ?? ""));
          if (!spotifyId) continue;
          const features: Partial<AudioFeatures> = {};
          let has = false;
          for (const key of FEATURE_KEYS) {
            if (item[key] != null) { features[key] = Number(item[key]); has = true; }
          }
          if (has) results.set(spotifyId, features as AudioFeatures);
        }
      }
    } catch {
      // Skip failed batches
    }

    onProgress?.(Math.min(i + BATCH_SIZE, trackIds.length), trackIds.length);
    if (i + BATCH_SIZE < trackIds.length) await sleep(RATE_LIMIT_MS);
  }

  return results;
}
